src/geeknn/ordination/_base.py

```diff
@@ -0,0 +1,282 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from itertools import chain
+from typing import Any
+
+import ee
+import numpy as np
+from joblib import Parallel, delayed
+from numpy.typing import NDArray
+from pydantic import BaseModel
+from sklearn.base import TransformerMixin
+
+from .utils import crosswalk_to_ids, filter_neighbors, return_k_neighbors, scores_to_fc
+
+
+class Geometry(BaseModel):
+    """Client-side proxy for ee.Geometry object."""
+
+    type: str
+    coordinates: list[Any]
+
+
+class Feature(BaseModel):
+    """Client-side proxy for ee.Feature object."""
+
+    type: str
+    geometry: Geometry
+    id: str
+    properties: dict[str, Any]
+
+    def get_property(self, property: str) -> Any:
+        return self.properties.get(property)
+
+    def get_properties(self, properties: list[str]) -> list[Any]:
+        return [self.get_property(x) for x in properties]
+
+
+class FeatureCollection(BaseModel):
+    """Client-side proxy for ee.FeatureCollection object."""
+
+    type: str
+    columns: dict[str, Any]
+    version: int
+    id: str
+    properties: dict[str, Any]
+    features: list[Feature]
+
+    def aggregate_array(self, property: str) -> NDArray:
+        """Return a single property as a 1-D numpy array."""
+        return np.array([f.get_property(property) for f in self.features])
+
+    def properties_to_array(self, properties: list[str]) -> NDArray:
+        """Return specified properties as a 2-D numpy array."""
+        return np.array([f.get_properties(properties) for f in self.features])
+
+    @classmethod
+    def from_ee_feature_collection(
+        cls, fc: ee.FeatureCollection, num_threads: int = -1, chunk_size: int = 5000
+    ):
+        """Create a client-side FeatureCollection from an ee.FeatureCollection
+        using multiple threads and chunking to avoid memory issues.
+        """
+        size = fc.size().getInfo()
+        info = fc.limit(0).getInfo()
+        chunks = [fc.toList(chunk_size, i) for i in range(0, size, chunk_size)]
+
+        with Parallel(n_jobs=num_threads, backend="threading") as p:
+            chunk_data = p(delayed(chunk.getInfo)() for chunk in chunks)
+
+        info.update({"features": list(chain.from_iterable(chunk_data))})
+        return cls(**info)
+
+
+class Raw:
+    def __init__(self, k=1, max_duplicates=None):
+        self.k = k
+        self.max_duplicates = max_duplicates if max_duplicates is not None else 5
```

Comment:
```suggestion
    def __init__(self, k=1, max_duplicates=5):
        self.k = k
        self.max_duplicates = max_duplicates
```

Was there a reason to avoid this?

Reply:
I think this was a carry-over from the original port from JS where I couldn't provide default parameters. Thanks for catching this!